multinode_integration_tests/src/main.rs

```diff
@@ -489,6 +592,125 @@ mod tests {
         assert_eq!(data_hunk.data, vec!(1, 2, 3, 4));
     }
 
+    #[test]
+    fn can_read_probe_target_from_str() {
+        probe_target_pairs()
+            .into_iter()
+            .for_each(|(string, expected)| {
+                let result = ProbeTarget::from_str(&string).unwrap();
+
+                assert_eq!(result, expected);
+            })
+    }
+
+    #[test]
+    fn display_implementation_works() {
+        probe_target_pairs()
+            .into_iter()
+            .for_each(|(expected, target)| {
+                let result = target.to_string();
+
+                assert_eq!(result, expected);
+            })
+    }
+
+    #[test]
+    fn handles_problems_reading_target_from_str() {
+        vec![
+            ("", "Syntax: <IP address>:<U|T><port>/<U|T><port>/..."),
+            (":", "Syntax: <IP address>:<U|T><port>/<U|T><port>/..."),
+            ("X:Y", "Syntax: <IP address>:<U|T><port>/<U|T><port>/..."),
+            (
+                "12.23.34.45:Y",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "X:U4321",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:4321",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:T",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:U4321/",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:U4321/5432",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:U4321/T",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
+            ),
+            (
+                "12.23.34.45:X4321",
+                "Syntax: <IP address>:<U|T><port>/<U|T><port>/...",
```

Comment:
Since the error response is always the same it is teasing to rewrite this so that the initial vector contains only the strings with the bad syntax, you can either hard-code the expected error message into the assertion in the `for_each()` or if you have the manifest constant for it by now, which I've already advised for its potentially handiness, you can simply put it in anyaway.

Reply:
Good suggestion.